### DISCUSSION TITLE
Berry: Google Graph in WebGUI

### DISCUSSION THREAD

**Discussion by ChrisVonFrosta** [arendst/Tasmota#21059] [Q&A]
i try to add google graph [example ](https://developers.google.com/chart/interactive/docs/gallery/areachart?hl=de)to the WebGUI.

```
def web_add_main_button()
  
    webserver.content_send(
      f"<script type='text/javascript' src='https://www.gstatic.com/charts/loader.js'></script>" +
      f"<script type='text/javascript'>" +
      f"google.charts.load('current', {{'packages':['corechart']}});" +
      f"google.charts.setOnLoadCallback(drawChart);" +
      f"function drawChart() {{" +
      f"var data = google.visualization.arrayToDataTable([" +
      f"['Year', 'Sales', 'Expenses']," +
      f"['2013',  1000,    400]," +
      f"['2014',  1170,    460]," +
      f"['2015',  660,    1120]," +
      f"['2016',  1030,    540]" +
      f"]);" +
      f"var options = {{" +
      f"title: 'Company Performance'," +
      f"hAxis: {{title: 'Year',  titleTextStyle: {{color: '#333'}}," +
      f"vAxis: {{minValue: 0}}" +
      f"}};" +
      f"var chart = new google.visualization.AreaChart(document.getElementById('chart_div'));" +
      f"chart.draw(data, options);" +
      f"}}" +
      f"<br>" +
      f"</script>" +
      
      f"<form action='/form_data' id='form_data'   method='post'>" + 
      f"<label for='sollwert'>Sollwert:</label><br>" +
      f"<input type='text' id='sollwert' name='sollwert' value='{persist.sollwert}'><br>" +
      f"<br>" +
      f"<label for='p_value'>P:</label><br>" +
      f"<input type='text' id='p_value' name='p_value' value='{persist.p_value}'><br>" +
      f"<label for='i_value'>I:</label><br>" +
      f"<input type='text' id='i_value' name='i_value' value='{persist.i_value}'><br>" +
      f"<label for='d_value'>D:</label><br>" +
      f"<input type='text' id='d_value' name='d_value' value='{persist.d_value}'><br>" +
      f"<br>" +
      f"<button name='form_data' class='button bgrn'>Save</button>" +
      f"</form>"
    )

end 
```

The form works but not the graph.
![grafik](https://github.com/arendst/Tasmota/assets/16272493/1e9ab989-a1d2-4b89-91c6-0222106f141b)

Any ideas?

**Comment by s-hadinger** [comment]
Did you check the html produced by this page? For ex with curl?

**Reply by sfromis** [reply]
Your browser "should" also have a debug console where detected errors will be visible.

(Your use of `{}` is not in balance, and you have html inside of the `<script>` block)

**Reply by ChrisVonFrosta** [reply]
thankyou so much!

```
def web_add_main_button()
  
    webserver.content_send(
      f"<script type='text/javascript' src='https://www.gstatic.com/charts/loader.js'></script>" +
      f"<script type='text/javascript'>" +
      f"google.charts.load('current', {{'packages':['corechart']}});" +
      f"google.charts.setOnLoadCallback(drawChart);" +
      f"function drawChart() {{" +
      f"var data = google.visualization.arrayToDataTable([" +
      f"['Year', 'Sales', 'Expenses']," +
      f"['2013',  1000,    400]," +
      f"['2014',  1170,    460]," +
      f"['2015',  660,    1120]," +
      f"['2016',  1030,    540]" +
      f"]);" +
      f"var options = {{" +
      f"title: 'Company Performance'," +
      f"hAxis: {{title: 'Year',  titleTextStyle: {{color: '#333'}}}}," +
      f"vAxis: {{minValue: 0}}" + 
      f"}};" +
      f"var chart = new google.visualization.AreaChart(document.getElementById('chart_div'));" +
      f"chart.draw(data, options);" +
      f"}}" +
      f"</script>" +

      f"<br>" +
      f"<div id='chart_div' style='width: 100%; height: 500px;'></div>" +
    )

end 
```

![grafik](https://github.com/arendst/Tasmota/assets/16272493/a95a17de-fbe8-43f6-8c83-961ce2a78789)

**Reply by barbudor** [reply]
There's no point in using f-string when you have no formatting to perform.
This is wasting CPU

**Reply by sfromis** [reply]
... and making it harder to read. You do not need those `+` at the end of each line, either.

When you come to the point of dynamically building the data to be charted, instead of fixed sample data, something like this could be useful (it is a list of lists):
```
var data=[]
data.push(['Year', 'Sales', 'Expenses'])
data.push(['2013', 1000, 400])
data.push(['2014', 1170, 460])
data.push(['2015', 660, 1120])
data.push(['2016', 1030, 540])
```
While this at first does not look better, it becomes easier to work with when you can build a variable number of rows one by one, and easily use vars instead of the constants.

And then something like:
`f"var data = google.visualization.arrayToDataTable({data});"`
This time, using a F-string is useful.

**Reply by ChrisVonFrosta** [reply]
you both are so right. It was very late and some new stuff.

I got it working with "live data".
I read one xiaomi mi thermometer.

I have a two scripts:

1.  `web.be` - that adds Web gui stuff and a trigger for updateing
2.  `pid_module.be` - that proceeds the sensor value, calculate the PID and PWM output and log the value to a dataset for the graph.

For having history data even after reboot, i safe the values to persistan variable.

MI32Period is 10. i react on each sensor value by saving it and adjusting a PID-Controller.
So i can see changes every 10 seconds (sensor update) in console. This runs as espected for minutes.
Then, it seems that script brokes the webgui or the Webgui/Graph kills something.
Nomore updates in console happend.
Then mainpage doesn't refresh normal. 
at fist refresh, the webgui is rendering until my custom stuff and nomore comes up (see pic 2). Sometimes nothing is loading (pic 3).
At this point, the (web) console is "freezed" (no more updates happend) but i can still restart via `restart 1` in web console.

Since i find no point for further informations i connected to console via usb (Tasmota Webinstaller -> view console).
This doesn't freeze and still showing updates from the `pir_module.be`.
Also updating values in `_persist.json` still happen.

So, somthing kills the "web enginge"?
if i close the main page, restart via console and first open the main page after 30 minutes, it starts normal. that it need just 3 Minutest

Pic1 (after reboot):
![grafik](https://github.com/arendst/Tasmota/assets/16272493/ced80ad3-c3da-40fe-9dbe-df5fa9385e17)

Pic 2 ( refresh start page after 3minutes
![grafik](https://github.com/arendst/Tasmota/assets/16272493/2c499154-b9b7-4374-a9aa-a3f14f9bdf21)

Pic 3 (Startpage down)
![grafik](https://github.com/arendst/Tasmota/assets/16272493/669a8564-34e1-4fb0-aadd-67ae957f16c5)


i append the script files as `.txt`. `.be`-files are not supported
[PID_humidy_bluetooth_module.txt](https://github.com/arendst/Tasmota/files/14814931/PID_humidy_bluetooth_module.txt)
[web_V2.txt](https://github.com/arendst/Tasmota/files/14814932/web_V2.txt)

**Reply by sfromis** [reply]
Seeing the low heap of 40 after 9 seconds of uptime makes me concerned that you may just be running out of RAM. BLE is already eating a lot there, and when you start to increase the load by other stuff. Of course, I can't say for sure that this is what is happening, but it would warrant a closer look.

More generally, "big" features on top of each other in the same device may be problematic, and you probably have no PSRAM even.

**Reply by ChrisVonFrosta** [reply]
>Seeing the low heap of 40 after 9 seconds..
could you explain that? what does  "low heap of 40" mean? i only see:
```
11:35:05:957   42
11:35:05:960   42
11:35:05:962   40
```

So, it could help to change bl sensor with i2c sensor and even has an ESP32 with PSRAM?
The Sensor thing is what i could test tonight. but it would be nice wiht bl sensors.

I have display boards wiht PSRAM. But than i would like to use the display too. what might compansate the extra PSRAM.

**Reply by ChrisVonFrosta** [reply]
maybe printing the memory status frequently could make sense?

**Reply by sfromis** [reply]
What I mean is unrelated to what you "only see" which looks like humidity data you are printing yourself.

I specifically talked about the Heap value from the STATE payload in one of your screenshots (not a good way of documenting logs, but at least a little bit was visible).

With the heap basically meaning free RAM, 40 KB is not much on ESP32 when you then start to do more stuff. And one thing is total free, but important is also fragmentation, which can preclude code wanting to allocate a buffer from working, even if total free was larger than the allocation.

The "Heap" value in state payloads does give a bit of info, and `tasmota.memory()` can give some details of the Berry side of memory management. Using `SetOption130` you can get heap/fragmentation status with every time stamp in the log.

While following the development can be "interesting", it does does not improve availability. I'd just be very careful about how much to add when heap is already that low, and strongly prefer a board including PSRAM, which typically is 2-16 MB of more RAM. While this extra RAM cannot be used for everything, Berry can take advantage of it.

**Comment by s-hadinger** [comment]
The way you build strings is a memory eater. You should have:

``` webserver.content_send(
      f"<script type='text/javascript' src='https://www.gstatic.com/charts/loader.js'></script>"
       "<script type='text/javascript'>"
       "google.charts.load('current', {{'packages':['corechart']}});"
       "google.charts.setOnLoadCallback(drawChart);"
       "function drawChart() {{"
       "var data = google.visualization.arrayToDataTable(["
       "['Year', 'Sales', 'Expenses'],"
       "['2013',  1000,    400],"
       "['2014',  1170,    460],"
       "['2015',  660,    1120],"
       "['2016',  1030,    540]"
       "]);"
       "var options = {{"
       "title: 'Company Performance',"
       "hAxis: {{title: 'Year',  titleTextStyle: {{color: '#333'}},"
       "vAxis: {{minValue: 0}}"
       "}};"
       "var chart = new google.visualization.AreaChart(document.getElementById('chart_div'));"
       "chart.draw(data, options);"
       "}}"
       "<br>"
       "</script>"
[...]
```

Only 1 f-string and remove `+` between strings. Otherwise it spends all its time building hundreds of temporary strings

**Reply by sfromis** [reply]
To explain - When you have multiple delimited strings adjacent, like this:
``` 
var v=123
print(f'a' 'b' 'c{v}' 'd')
abc123d
```
Those strings are parsed as one long string during compilation, hence the `f` does not belong to a single set of string delimiters but is for the whole resulting string. That's why I already pointed out not to have those `+` at the end of each of your lines, as this is a runtime string concatenation instead of already having one long string.

**Reply by ChrisVonFrosta** [reply]
thank you both for pointing again and explaining!
the behaviar with only one `f"string"` is much better. Until now, (re)loading the start page doesn't crash the gui. the first of two refreshes doesn't load the graph, but the next one does.

**Comment by ChrisVonFrosta** [comment]
since you told deactivation of BL save a lot of memory. Does that mean that using an i2c sensor would use much less memory?

**Reply by sfromis** [reply]
If BLE is what you mean, yes a build without Bluetooth support does avoid a good chunk of memory usage. I'm seeing a difference of over 90 KB.

**Reply by ChrisVonFrosta** [reply]
yes, BLE.
i need a build without Bluetooth support? deactivating in in web config is helps not?

**Reply by sfromis** [reply]
Right. Ticking it off in "Configure BLE" still eats more RAM than not having it. And for an I2C sensor, you probably want another build anyway, as the standard Bluetooth builds does not include I2C sensor drivers.

**Reply by ChrisVonFrosta** [reply]
if i use a board with esp32 with PSRAM, i have to use `Tasmota32-S3 (QIO Flash/OPI PSRAM)`. i think there is no bluetooth available? do i need to make a custom build if for bluettooth and PSRAM?

**Reply by sfromis** [reply]
While some ESP32-S3 boards with PSRAM do use octal SPI (OPI), this is different per board, and others don't. And PSRAM is not only with ESP32-S3.

Here's one example of a small dev board with ESP32-S3 and PSRAM without using octal SPI:
https://templates.blakadder.com/unexpectedmaker_tinys3.html

**Reply by ChrisVonFrosta** [reply]
thx. but the only prebuild image with Bluetooth is the tasmota32-bluettooth binary?

**Reply by sfromis** [reply]
With the socalled ["unofficial" development builds](https://github.com/tasmota/install/tree/firmware/firmware/unofficial), there are also Bluetooth-specific binaries for ESP32-S3 and ESP32-C3.

**Reply by ChrisVonFrosta** [reply]
thx.
i think about. the module (pid_module.be) writes the history data in the persist variable. the script (web.de) reads that vars when webpage ist (re)loading.
Could that make problems (race condition)? i write the whole list "datenliste" in the persist var every time. Actually max list length is 200

**Reply by ChrisVonFrosta** [reply]
for testing, i commented out the BLEAlias' in autoexec.bat.
then the main page reloads every time, very fast!
so it may depends with the way i stored the data?!

heap_free is increased with no binded BLEAlias.
from 25-35 before to ~45 without BLEAlias.

**Reply by sfromis** [reply]
Well, did you check how free heap is going?

Anyway, your technique of saving `persist` every time you've done an update is on the extreme side, and contributes to memory thrashing. It will have to build a long string, probably above 10 KB, and save it to the file. It is also a factor in wearing down flash storage. With the paucity of RAM, you may see better results without doing that `persist.save()` all the time. This saving is not needed for the web server code to read the data.

**Reply by ChrisVonFrosta** [reply]
>Well, did you check how free heap is going?

i just wrote thus before.

>heap_free is increased with no binded BLEAlias.
from 25-35 before to ~45 without BLEAlias.

>This saving is not needed for the web server code to read the data.
i do `persist.save()` on each cycle to have a full dataset after reboot. I won't like to loose the history if rebooting.

after reactivate only one BLEAlias (before it were two) it still works smooth. Webpage refresh super fast. I don't know why.
Maybe the second BLE sensor made the difference?!

no recently refresh:
```
tasmota.memory()
`{'heap_free': 42, 'program_free': 1244, 'program': 1635, 'stack_low': 3.53125, 'iram_free': 2, 'frag': 53, 'flash': 4096, 'stack_size': 8, 'flash_real': 4096}
```
recently refreshed:
```
tasmota.memory()
{'heap_free': 30, 'program_free': 1244, 'program': 1635, 'stack_low': 3.53125, 'iram_free': 2, 'frag': 75, 'flash': 4096, 'stack_size': 8, 'flash_real': 4096}
```

**Reply by sfromis** [reply]
Heap going down from 42 to 30 is pretty bad when you are pushing that large strings around by all those `persist.save()`. You are pushing it way too hard.

**Comment by ChrisVonFrosta** [comment]
oh damn.. i came back after 2 hours. The script still rans but after reloading websitenothing came up again, even the console in extra window stops ;/

**Reply by sfromis** [reply]
Well, if you have the console, you can check if Heap was counting down...

**Comment by ChrisVonFrosta** [comment]
Maybe i could ask you two more things related to this.

1. is there an option periodically reload the chart ?
2. This BLE thermometer displays its humidity value in the web GUI with one decimal place. However, in the telemessage, where I grab the value, it comes without decimals. Where do i loose the decimal?

```
#module file
def pars_humidity(value, trigger, msg)
    var humid=   real(value['Humidity'])
    print(value)
    print(real(humid))
    update_pid(humid)
end

 pid_module.pars_humidity =   def (value, trigger, msg)
    pars_humidity(value, trigger, msg)
end

# other file
tasmota.add_rule("ATCf466a9", pm.pars_humidity)
```

results in this ouput:
```
01:58:23.815 MQT: tele/tasmota_D503A8/SENSOR = {"Time":"2024-04-01T01:58:23","ATCf466a9":{"alias":"Innnen","mac":"a4c138f466a9","Temperature":22.1,"Humidity":56.0,"DewPoint":12.9,"Battery":94,"RSSI":-48},"TempUnit":"C"}
01:58:23.836 {'mac': 'a4c138f466a9', 'RSSI': -48, 'Humidity': 56, 'alias': 'Innnen', 'Temperature': 22.1, 'Battery': 94, 'DewPoint': 12.9}
```

**Reply by sfromis** [reply]
The way Tasmota handles refreshing the front page sensor values is by not refreshing it, but by having JavaScript code in the browser do polling instead.

To see where you are loosing the decimals, look in your own code, like you choosing to use the `int()` function. What you posted now is not losing any decimals, just a different way of formatting a number which has no decimals.

More generally, measuring humidity is already so inaccurate that decimals is a fiction.

**Reply by ChrisVonFrosta** [reply]
i am confused. What i posted now, is still loosing the decimal. 

i noticed that i faulty used the `int()` functions before. It was used to convert strings to numbers.
That's why i tried with `real()` now.


EDIT: My Fault, i missed to post the output from `print(real(humid))` before
but it was `56` without decimals

**Reply by sfromis** [reply]
As I already said, you are **not** losing any decimals, the number 56 does not become any different by being written as 56.0. It is still the number 56 which does not have any decimals.

Note that it is a waste of typing to use `real()` (or `int()` for that matter) function here:
`var humid=   real(value['Humidity'])`
What you have in the parsed map is already a number, not a string to be parsed again.

**Reply by barbudor** [reply]
Your tele message does have the decimal on humidity:
```
01:58:23.815 MQT: tele/tasmota_D503A8/SENSOR = {"Time":"2024-04-01T01:58:23","ATCf466a9":{"alias":"Innnen","mac":"a4c138f466a9","Temperature":22.1,"Humidity":56.0,"DewPoint":12.9,"Battery":94,"RSSI":-48},"TempUnit":"C"}
```
As far as I can see in the code, the publication is done based on `HumRes` command

As fars as I understand the line below is a Berry print of message map so there is nothing strange as 56.0 is printed as 56 as you can test yourself in the console
![image](https://github.com/arendst/Tasmota/assets/2996042/6f3b30b9-8236-48b5-a249-593b2ebceb08)

**Reply by sfromis** [reply]
If you want the number 56, which is exactly the same as when formatting it as 56.0, expressed as a string with one digit after a decimal point...
```
var hum=56
print(f'{hum:.1f}')
```

**Reply by ChrisVonFrosta** [reply]
omg.... i only saw the decimals in the web gui and never noticed that this is allways zero.
![grafik](https://github.com/arendst/Tasmota/assets/16272493/a17a7897-b955-4445-9bd7-267b1901e094)


And i didn't use my brain. This thermoneter have an accuracy or mean error from 1-2% for humidity...

**Reply by sfromis** [reply]
As barbudor already mentioned, the number of decimals on Humidity is up to what you configure in `HumRes`, not what value the sensor actually tries to deliver. And the default is indeed 1, no matter how meaningless decimal digits may be.
